tests/test-class-wc-payment-gateway-wcpay-subscriptions-process-payment.php

```diff
@@ -337,6 +337,47 @@ public function test_saved_card_is_added_before_status_update() {
 		$result = $this->mock_wcpay_gateway->process_payment( $order->get_id() );
 	}
 
+	public function test_card_is_saved_when_updating_subscription_payment_method() {
+		$order = WC_Helper_Order::create_order( self::USER_ID, 0 );
+
+		$_GET = [ 'change_payment_method' => 10 ];
+
+		$this->mock_wcs_order_contains_subscription( false );
+
+		WCS_Mock::set_wcs_is_subscription(
+			function ( $order ) {
+				return true;
+			}
+		);
+
+		$this->mock_api_client
+			->expects( $this->once() )
+			->method( 'create_setup_intent' )
+			->with( self::PAYMENT_METHOD_ID, self::CUSTOMER_ID, 'true' )
+			->willReturn( $this->setup_intent );
+
+		$this->mock_token_service
+			->expects( $this->once() )
+			->method( 'add_payment_method_to_user' )
+			->with( self::PAYMENT_METHOD_ID, $order->get_user() )
+			->willReturn( $this->token );
+
+		// Expect add token to order to be called, so it can be reused in renewals.
+		$this->mock_wcpay_gateway
+			->expects( $this->once() )
+			->method( 'add_token_to_order' )
+			->with(
+				$this->callback( $this->match_order_id( $order->get_id() ) ),
+				$this->token
+			);
```

Comment:
I've played with it and realized we'll have to fix mocking of `add_token_to_order` in all other tests which is out of scope of the PR. So let's move it into own issue (`Good first issue`) if it makes sense to refactor at all.

Reply:
Yeah, it is a bit strange to mock one of the class's methods to test another one. The main reason for this was to split the tests for `add_token_to_order` and test it separately from the checkout flow, as it has a few edge cases. This mock allows the `process_payment` tests to avoid testing the add token logic, but rather check if it was called.

There's also the case of the `test_saved_card_is_added_before_status_update` test. We need to assert that the token was added before the order status is updated (more details in https://github.com/Automattic/woocommerce-payments/pull/832/commits/2dfec7df497014a6cf0582315821a8114fb9bd32). I'm not sure how to assert that if we're not mocking it unless we hook into the `woocommerce_order_status_changed` and check if the token is already there.

What are your thoughts on this? In case you think this partial mock is not doing us any favors, I'm fine with removing it and testing the basic `add_token_to_order` case on those tests, as mocking it does seem like a bit of a code smell.

Reply:
Thanks for sharing your thoughts @luizreis . I see your point, but it's hard to comment on edge cases you've metnioned without deeper dive into it. I've just noticed from the first look that the code doesn't look like a candidate for mocking. I might be completely wrong here though 🙂

> What are your thoughts on this? In case you think this partial mock is not doing us any favors, I'm fine with removing it and testing the basic add_token_to_order case on those tests, as mocking it does seem like a bit of a code smell.

As I said above IMO it doesn't worth to refactor it as part of this PR. Creating an issue for that with comments above as context would be enough IMO, but the final decision is yours ;)

Reply:
I think it's worth considering, and I agree that it doesn't belong to this PR. So, I created https://github.com/Automattic/woocommerce-payments/issues/920 for us to consider it and possibly remove it in the future.

Thank you for suggesting this!